// @liquidmetal-ai/raindrop-framework/dist/core/cors.js
/**
 * CORS (Cross-Origin Resource Sharing) configuration and handling utilities
 * for Raindrop framework applications.
 */
/**
 * Matches the request origin against the CORS configuration
 * @param request The incoming request
 * @param env The environment object with bindings
 * @param config CORS configuration
 * @returns The allowed origin value or null if not allowed
 */
export const matchOrigin = (request, env, config) => {
    const requestOrigin = request.headers.get('origin');
    if (!requestOrigin) {
        return null;
    }
    const { origin } = config;
    if (origin === '*') {
        return '*';
    }
    if (typeof origin === 'function') {
        return origin(request, env);
    }
    if (typeof origin === 'string') {
        return requestOrigin === origin ? origin : null;
    }
    if (Array.isArray(origin)) {
        return origin.includes(requestOrigin) ? requestOrigin : null;
    }
    return null;
};
/**
 * Adds CORS headers to an existing response
 * @param response The response to add headers to
 * @param request The incoming request
 * @param env The environment object with bindings
 * @param config CORS configuration
 * @returns A new Response with CORS headers added
 */
export const addCorsHeaders = (response, request, env, config) => {
    const allowedOrigin = matchOrigin(request, env, config);
    if (!allowedOrigin) {
        // Origin not allowed, return response unchanged
        return response;
    }
    const headers = new Headers(response.headers);
    headers.set('Access-Control-Allow-Origin', allowedOrigin);
    if (config.credentials) {
        headers.set('Access-Control-Allow-Credentials', 'true');
    }
    if (config.exposeHeaders && config.exposeHeaders.length > 0) {
        headers.set('Access-Control-Expose-Headers', config.exposeHeaders.join(', '));
    }
    // Vary header tells caches that the response varies based on Origin
    const vary = headers.get('Vary');
    if (vary) {
        if (!vary.includes('Origin')) {
            headers.set('Vary', `${vary}, Origin`);
        }
    }
    else {
        headers.set('Vary', 'Origin');
    }
    return new Response(response.body, {
        status: response.status,
        statusText: response.statusText,
        headers
    });
};
/**
 * Handles CORS preflight (OPTIONS) requests
 * @param request The incoming OPTIONS request
 * @param env The environment object with bindings
 * @param config CORS configuration
 * @returns A Response for the preflight request
 */
export const handlePreflight = (request, env, config) => {
    const allowedOrigin = matchOrigin(request, env, config);
    if (!allowedOrigin) {
        // Origin not allowed, return 403
        return new Response(null, { status: 403 });
    }
    const headers = new Headers();
    headers.set('Access-Control-Allow-Origin', allowedOrigin);
    if (config.credentials) {
        headers.set('Access-Control-Allow-Credentials', 'true');
    }
    const allowMethods = config.allowMethods || ['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'OPTIONS'];
    headers.set('Access-Control-Allow-Methods', allowMethods.join(', '));
    const allowHeaders = config.allowHeaders || ['Content-Type', 'Authorization'];
    headers.set('Access-Control-Allow-Headers', allowHeaders.join(', '));
    const maxAge = config.maxAge ?? 86400; // 24 hours default
    headers.set('Access-Control-Max-Age', maxAge.toString());
    headers.set('Vary', 'Origin');
    return new Response(null, {
        status: 204,
        headers
    });
};
/**
 * Creates a CORS handler function from a configuration object.
 * This is the main factory function that users will call.
 *
 * @param config CORS configuration
 * @returns A CORS handler function
 *
 * @example
 * ```typescript
 * export const cors = createCorsHandler({
 *   origin: ['https://example.com', 'https://app.example.com'],
 *   allowMethods: ['GET', 'POST'],
 *   credentials: true
 * });
 * ```
 */
export const createCorsHandler = (config) => {
    return (request, env, response) => {
        // If no response provided, this is a preflight OPTIONS request
        if (!response) {
            return handlePreflight(request, env, config);
        }
        // Otherwise, add CORS headers to the existing response
        return addCorsHeaders(response, request, env, config);
    };
};
/**
 * Pre-configured CORS handler that allows all origins.
 * Useful for public APIs and development.
 *
 * @example
 * ```typescript
 * export { corsAllowAll as cors } from '@liquidmetal-ai/raindrop-framework/core/cors';
 * ```
 */
export const corsAllowAll = createCorsHandler({
    origin: '*',
    allowMethods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'OPTIONS'],
    allowHeaders: ['Content-Type', 'Authorization'],
    credentials: false
});
/**
 * Default CORS handler that disables CORS by not adding any CORS headers.
 * This is the secure default - requests from browsers will be blocked by CORS policy.
 *
 * Preflight OPTIONS requests return 403 Forbidden.
 * Other requests pass through without CORS headers, so browsers will block cross-origin access.
 *
 * To enable CORS, replace this with:
 * - `corsAllowAll` for public APIs that allow any origin
 * - `createCorsHandler({ origin: [...] })` to allow specific origins
 * - A custom handler function for advanced control
 *
 * @example
 * ```typescript
 * // Secure default - no CORS headers
 * export { corsDisabled as cors } from '@liquidmetal-ai/raindrop-framework/core/cors';
 *
 * // Or enable CORS for specific origins:
 * export const cors = createCorsHandler({
 *   origin: ['https://example.com'],
 *   credentials: true
 * });
 * ```
 */
export const corsDisabled = (request, _env, response) => {
    // Handle OPTIONS preflight requests
    if (!response && request.method === 'OPTIONS') {
        // Return 403 to reject preflight requests
        return new Response(null, { status: 403 });
    }
    // Defensive check - response should always be provided for non-OPTIONS requests
    if (!response) {
        throw new Error('corsDisabled called without response for non-OPTIONS request');
    }
    // Return response unchanged (no CORS headers)
    return response;
};
