// @liquidmetal-ai/raindrop-framework/dist/ai.js
/* eslint-disable */
/** Complete catalog of all available AI models */
export const MODEL_CATALOG = [
    {
        name: 'llama-3.3-70b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'llama-3.3-70b',
        description: 'External model routed through model-router: llama-3.3-70b'
    },
    {
        name: 'llama-4-maverick-17b',
        capability: ['chat', 'vision'],
        provider: 'external',
        providerModel: 'llama-4-maverick-17b',
        description: 'External model routed through model-router: llama-4-maverick-17b'
    },
    {
        name: 'llama-3.1-8b-external',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'llama-3.1-8b-instruct',
        description: 'External model routed through model-router: llama-3.1-8b-instruct'
    },
    {
        name: 'deepseek-r1',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'deepseek-r1',
        description: 'External model routed through model-router: deepseek-r1'
    },
    {
        name: 'deepseek-v3-0324',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'deepseek-v3-0324',
        description: 'External model routed through model-router: deepseek-v3-0324'
    },
    {
        name: 'deepseek-v3.1',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'deepseek-v3.1',
        description: 'External model routed through model-router: deepseek-v3.1'
    },
    {
        name: 'deepseek-v3.1-terminus',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'deepseek-v3.1-terminus',
        description: 'External model routed through model-router: deepseek-v3.1-terminus'
    },
    {
        name: 'deepseek-r1-distill-llama-70b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'deepseek-r1-distill-llama-70b',
        description: 'External model routed through model-router: deepseek-r1-distill-llama-70b'
    },
    {
        name: 'qwen-3-32b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'qwen-3-32b',
        description: 'External model routed through model-router: qwen-3-32b'
    },
    {
        name: 'llama-3.3-swallow-70b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'llama-3.3-swallow-70b',
        description: 'External model routed through model-router: llama-3.3-swallow-70b'
    },
    {
        name: 'whisper-large-v3',
        capability: ['audio'],
        provider: 'external',
        providerModel: 'whisper-large-v3',
        description: 'External model routed through model-router: whisper-large-v3'
    },
    {
        name: 'gpt-oss-120b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'gpt-oss-120b',
        description: 'External model routed through model-router: gpt-oss-120b'
    },
    {
        name: 'llama-3.1-8b-instant',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'llama-3.1-8b-instant',
        description: 'External model routed through model-router: llama-3.1-8b-instant'
    },
    {
        name: 'gemma-9b-it',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'gemma-9b-it',
        description: 'External model routed through model-router: gemma-9b-it'
    },
    {
        name: 'whisper-large-v3-turbo',
        capability: ['audio'],
        provider: 'external',
        providerModel: 'whisper-large-v3-turbo',
        description: 'External model routed through model-router: whisper-large-v3-turbo'
    },
    {
        name: 'llama-guard-4-12b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'llama-guard-4-12b',
        description: 'External model routed through model-router: llama-guard-4-12b'
    },
    {
        name: 'kimi-k2',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'kimi-k2',
        description: 'External model routed through model-router: kimi-k2'
    },
    {
        name: 'gpt-oss-20b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'gpt-oss-20b',
        description: 'External model routed through model-router: gpt-oss-20b'
    },
    {
        name: 'embeddings',
        capability: ['embeddings'],
        provider: 'external',
        providerModel: 'bge-embeddings',
        description: 'External model routed through model-router: bge-embeddings'
    },
    {
        name: 'pii-detection',
        capability: ['pii-detection'],
        provider: 'external',
        providerModel: 'piiranha-pii',
        description: 'External model routed through model-router: piiranha-pii'
    },
    {
        name: 'qwen-3-coder-480b',
        capability: ['chat'],
        provider: 'external',
        providerModel: 'qwen-3-coder-480b',
        description: 'External model routed through model-router: qwen-3-coder-480b'
    },
    {
        name: 'smart-turn-v2',
        capability: ['audio'],
        provider: 'cloudflare',
        providerModel: '@cf/pipecat-ai/smart-turn-v2',
        description: 'An open source, community-driven, native audio turn detection model in 2nd version'
    },
    {
        name: 'qwen-1.5-0.5b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/qwen/qwen1.5-0.5b-chat',
        description: 'Qwen1.5 is the improved version of Qwen, the large language model series developed by Alibaba Cloud.'
    },
    {
        name: 'bge-m3',
        capability: ['embeddings'],
        provider: 'cloudflare',
        providerModel: '@cf/baai/bge-m3',
        description: 'Multi-Functionality, Multi-Linguality, and Multi-Granularity embeddings model.'
    },
    {
        name: 'distilbert-sst-2-int8',
        capability: ['text-classification'],
        provider: 'cloudflare',
        providerModel: '@cf/huggingface/distilbert-sst-2-int8',
        description: 'Distilled BERT model that was finetuned on SST-2 for sentiment classification'
    },
    {
        name: 'gemma-2b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/google/gemma-2b-it-lora',
        description: 'This is a Gemma-2B base model that Cloudflare dedicates for inference with LoRA adapters. Gemma is a family of lightweight, state-of-the-art open models from Google, built from the same research and technology used to create the Gemini models.'
    },
    {
        name: 'starling-lm-7b-beta',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/nexusflow/starling-lm-7b-beta',
        description: 'We introduce Starling-LM-7B-beta, an open large language model (LLM) trained by Reinforcement Learning from AI Feedback (RLAIF). Starling-LM-7B-beta is trained from Openchat-3.5-0106 with our new reward model Nexusflow/Starling-RM-34B and policy optimization method Fine-Tuning Language Models from Human Preferences (PPO).'
    },
    {
        name: 'llama-3-8b-instruct',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3-8b-instruct',
        description: 'Generation over generation, Meta Llama 3 demonstrates state-of-the-art performance on a wide range of industry benchmarks and offers new capabilities, including improved reasoning.'
    },
    {
        name: 'llama-3.2-3b-instruct',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.2-3b-instruct',
        description: 'The Llama 3.2 instruction-tuned text only models are optimized for multilingual dialogue use cases, including agentic retrieval and summarization tasks.'
    },
    {
        name: 'llamaguard-7b-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/llamaguard-7b-awq',
        description: 'Llama Guard is a model for classifying the safety of LLM prompts and responses, using a taxonomy of safety risks. '
    },
    {
        name: 'neural-chat-7b-v3-1-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/neural-chat-7b-v3-1-awq',
        description: 'This model is a fine-tuned 7B parameter LLM on the Intel Gaudi 2 processor from the mistralai/Mistral-7B-v0.1 on the open source dataset Open-Orca/SlimOrca.'
    },
    {
        name: 'llama-guard-3-8b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-guard-3-8b',
        description: 'Llama Guard 3 is a Llama-3.1-8B pretrained model, fine-tuned for content safety classification. Similar to previous versions, it can be used to classify content in both LLM inputs (prompt classification) and in LLM responses (response classification). It acts as an LLM – it generates text in its output that indicates whether a given prompt or response is safe or unsafe, and if unsafe, it also lists the content categories violated.'
    },
    {
        name: 'llama-2-7b-chat-fp16',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-2-7b-chat-fp16',
        description: 'Full precision (fp16) generative text model with 7 billion parameters from Meta'
    },
    {
        name: 'mistral-7b-instruct',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/mistral/mistral-7b-instruct-v0.1',
        description: 'Instruct fine-tuned version of the Mistral-7b generative text model with 7 billion parameters'
    },
    {
        name: 'melotts',
        capability: ['tts'],
        provider: 'cloudflare',
        providerModel: '@cf/myshell-ai/melotts',
        description: 'MeloTTS is a high-quality multi-lingual text-to-speech library by MyShell.ai.'
    },
    {
        name: 'mistral-7b-instruct-v0.2-lora',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/mistral/mistral-7b-instruct-v0.2-lora',
        description: 'The Mistral-7B-Instruct-v0.2 Large Language Model (LLM) is an instruct fine-tuned version of the Mistral-7B-v0.2.'
    },
    {
        name: 'aura-2-es',
        capability: ['tts'],
        provider: 'cloudflare',
        providerModel: '@cf/deepgram/aura-2-es',
        description: 'Aura-2 is a context-aware text-to-speech (TTS) model that applies natural pacing, expressiveness, and fillers based on the context of the provided text. The quality of your text input directly impacts the naturalness of the audio output.'
    },
    {
        name: 'whisper',
        capability: ['audio'],
        provider: 'cloudflare',
        providerModel: '@cf/openai/whisper',
        description: 'Whisper is a general-purpose speech recognition model. It is trained on a large dataset of diverse audio and is also a multitasking model that can perform multilingual speech recognition, speech translation, and language identification.'
    },
    {
        name: 'tinyllama-1.1b-chat-v1.0',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/tinyllama/tinyllama-1.1b-chat-v1.0',
        description: 'The TinyLlama project aims to pretrain a 1.1B Llama model on 3 trillion tokens. This is the chat model finetuned on top of TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T.'
    },
    {
        name: 'plamo-embedding-1b',
        capability: ['embeddings'],
        provider: 'cloudflare',
        providerModel: '@cf/pfnet/plamo-embedding-1b',
        description: 'PLaMo-Embedding-1B is a Japanese text embedding model developed by Preferred Networks, Inc.  It can convert Japanese text input into numerical vectors and can be used for a wide range of applications, including information retrieval, text classification, and clustering.'
    },
    {
        name: 'mistral-7b-instruct-v0.2',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/mistral/mistral-7b-instruct-v0.2',
        description: 'The Mistral-7B-Instruct-v0.2 Large Language Model (LLM) is an instruct fine-tuned version of the Mistral-7B-v0.2. Mistral-7B-v0.2 has the following changes compared to Mistral-7B-v0.1: 32k context window (vs 8k context in v0.1), rope-theta = 1e6, and no Sliding-Window Attention.'
    },
    {
        name: 'una-cybertron-7b-v2-bf16',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/fblgit/una-cybertron-7b-v2-bf16',
        description: 'Cybertron 7B v2 is a 7B MistralAI based model, best on it\'s series. It was trained with SFT, DPO and UNA (Unified Neural Alignment) on multiple datasets.'
    },
    {
        name: 'llava-1.5-7b',
        capability: ['vision', 'chat'],
        provider: 'cloudflare',
        providerModel: '@cf/llava-hf/llava-1.5-7b-hf',
        description: 'LLaVA is an open-source chatbot trained by fine-tuning LLaMA/Vicuna on GPT-generated multimodal instruction-following data. It is an auto-regressive language model, based on the transformer architecture.'
    },
    {
        name: 'deepseek-r1-distill-qwen-32b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/deepseek-ai/deepseek-r1-distill-qwen-32b',
        description: 'DeepSeek-R1-Distill-Qwen-32B is a model distilled from DeepSeek-R1 based on Qwen2.5. It outperforms OpenAI-o1-mini across various benchmarks, achieving new state-of-the-art results for dense models.'
    },
    {
        name: 'stable-diffusion-v1-5-inpainting',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/runwayml/stable-diffusion-v1-5-inpainting',
        description: 'Stable Diffusion Inpainting is a latent text-to-image diffusion model capable of generating photo-realistic images given any text input, with the extra capability of inpainting the pictures by using a mask.'
    },
    {
        name: 'flux',
        capability: ['audio'],
        provider: 'cloudflare',
        providerModel: '@cf/deepgram/flux',
        description: 'Flux is the first conversational speech recognition model built specifically for voice agents.'
    },
    {
        name: 'nova-3',
        capability: ['audio'],
        provider: 'cloudflare',
        providerModel: '@cf/deepgram/nova-3',
        description: 'Transcribe audio using Deepgram’s speech-to-text model'
    },
    {
        name: 'flux-1-schnell',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/black-forest-labs/flux-1-schnell',
        description: 'FLUX.1 [schnell] is a 12 billion parameter rectified flow transformer capable of generating images from text descriptions. '
    },
    {
        name: 'discolm-german-7b-v1-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/thebloke/discolm-german-7b-v1-awq',
        description: 'DiscoLM German 7b is a Mistral-based large language model with a focus on German-language applications. AWQ is an efficient, accurate and blazing-fast low-bit weight quantization method, currently supporting 4-bit quantization.'
    },
    {
        name: 'llama-2-7b-chat-int8',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-2-7b-chat-int8',
        description: 'Quantized (int8) generative text model with 7 billion parameters from Meta'
    },
    {
        name: 'llama-3.1-8b-instruct-fp8',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.1-8b-instruct-fp8',
        description: 'Llama 3.1 8B quantized to FP8 precision'
    },
    {
        name: 'mistral-7b-instruct-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/mistral-7b-instruct-v0.1-awq',
        description: 'Mistral 7B Instruct v0.1 AWQ is an efficient, accurate and blazing-fast low-bit weight quantized Mistral variant.'
    },
    {
        name: 'qwen-1.5-7b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/qwen/qwen1.5-7b-chat-awq',
        description: 'Qwen1.5 is the improved version of Qwen, the large language model series developed by Alibaba Cloud. AWQ is an efficient, accurate and blazing-fast low-bit weight quantization method, currently supporting 4-bit quantization.'
    },
    {
        name: 'llama-3.2-1b-instruct',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.2-1b-instruct',
        description: 'The Llama 3.2 instruction-tuned text only models are optimized for multilingual dialogue use cases, including agentic retrieval and summarization tasks.'
    },
    {
        name: 'llama-2-13b-chat-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/llama-2-13b-chat-awq',
        description: 'Llama 2 13B Chat AWQ is an efficient, accurate and blazing-fast low-bit weight quantized Llama 2 variant.'
    },
    {
        name: 'resnet-50',
        capability: ['image-classification'],
        provider: 'cloudflare',
        providerModel: '@cf/microsoft/resnet-50',
        description: '50 layers deep image classification CNN trained on more than 1M images from ImageNet'
    },
    {
        name: 'stable-diffusion-xl-lightning',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/bytedance/stable-diffusion-xl-lightning',
        description: 'SDXL-Lightning is a lightning-fast text-to-image generation model. It can generate high-quality 1024px images in a few steps.'
    },
    {
        name: 'deepseek-coder-6.7b-base',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/deepseek-coder-6.7b-base-awq',
        description: 'Deepseek Coder is composed of a series of code language models, each trained from scratch on 2T tokens, with a composition of 87% code and 13% natural language in both English and Chinese.'
    },
    {
        name: 'llama-2-7b-chat-hf-lora',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta-llama/llama-2-7b-chat-hf-lora',
        description: 'This is a Llama2 base model that Cloudflare dedicated for inference with LoRA adapters. Llama 2 is a collection of pretrained and fine-tuned generative text models ranging in scale from 7 billion to 70 billion parameters. This is the repository for the 7B fine-tuned model, optimized for dialogue use cases and converted for the Hugging Face Transformers format. '
    },
    {
        name: 'llama-3.3-70b-instruct-fp8',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.3-70b-instruct-fp8-fast',
        description: 'Llama 3.3 70B quantized to fp8 precision, optimized to be faster.'
    },
    {
        name: 'granite-4.0-h-micro',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/ibm-granite/granite-4.0-h-micro',
        description: 'Granite 4.0 instruct models deliver strong performance across benchmarks, achieving industry-leading results in key agentic tasks like instruction following and function calling. These efficiencies make the models well-suited for a wide range of use cases like retrieval-augmented generation (RAG), multi-agent workflows, and edge deployments.'
    },
    {
        name: 'dreamshaper-8-lcm',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/lykon/dreamshaper-8-lcm',
        description: 'Stable Diffusion model that has been fine-tuned to be better at photorealism without sacrificing range.'
    },
    {
        name: 'phoenix-1.0',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/leonardo/phoenix-1.0',
        description: 'Phoenix 1.0 is a model by Leonardo.Ai that generates images with exceptional prompt adherence and coherent text.'
    },
    {
        name: 'stable-diffusion-xl-base-1.0',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/stabilityai/stable-diffusion-xl-base-1.0',
        description: 'Diffusion-based text-to-image generative model by Stability AI. Generates and modify images based on text prompts.'
    },
    {
        name: 'openhermes-2.5-mistral-7b-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/openhermes-2.5-mistral-7b-awq',
        description: 'OpenHermes 2.5 Mistral 7B is a state of the art Mistral Fine-tune, a continuation of OpenHermes 2 model, which trained on additional code datasets.'
    },
    {
        name: 'm2m100-1.2b',
        capability: ['translation'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/m2m100-1.2b',
        description: 'Multilingual encoder-decoder (seq-to-seq) model trained for Many-to-Many multilingual translation'
    },
    {
        name: 'indictrans2-en-indic-1B',
        capability: ['translation'],
        provider: 'cloudflare',
        providerModel: '@cf/ai4bharat/indictrans2-en-indic-1B',
        description: 'IndicTrans2 is the first open-source transformer-based multilingual NMT model that supports high-quality translations across all the 22 scheduled Indic languages'
    },
    {
        name: 'deepseek-coder-6.7b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/deepseek-coder-6.7b-instruct-awq',
        description: 'Deepseek Coder is composed of a series of code language models, each trained from scratch on 2T tokens, with a composition of 87% code and 13% natural language in both English and Chinese.'
    },
    {
        name: 'bge-small-en',
        capability: ['embeddings'],
        provider: 'cloudflare',
        providerModel: '@cf/baai/bge-small-en-v1.5',
        description: 'BAAI general embedding (Small) model that transforms any given text into a 384-dimensional vector'
    },
    {
        name: 'qwen-coder-32b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/qwen/qwen2.5-coder-32b-instruct',
        description: 'Qwen2.5-Coder is the latest series of Code-Specific Qwen large language models (formerly known as CodeQwen). As of now, Qwen2.5-Coder has covered six mainstream model sizes, 0.5, 1.5, 3, 7, 14, 32 billion parameters, to meet the needs of different developers. Qwen2.5-Coder brings the following improvements upon CodeQwen1.5:'
    },
    {
        name: 'deepseek-math-7b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/deepseek-ai/deepseek-math-7b-instruct',
        description: 'DeepSeekMath-Instruct 7B is a mathematically instructed tuning model derived from DeepSeekMath-Base 7B. DeepSeekMath is initialized with DeepSeek-Coder-v1.5 7B and continues pre-training on math-related tokens sourced from Common Crawl, together with natural language and code data for 500B tokens.'
    },
    {
        name: 'falcon-7b-instruct',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/tiiuae/falcon-7b-instruct',
        description: 'Falcon-7B-Instruct is a 7B parameters causal decoder-only model built by TII based on Falcon-7B and finetuned on a mixture of chat/instruct datasets.'
    },
    {
        name: 'hermes-2-pro-mistral-7b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/nousresearch/hermes-2-pro-mistral-7b',
        description: 'Hermes 2 Pro on Mistral 7B is the new flagship 7B Hermes! Hermes 2 Pro is an upgraded, retrained version of Nous Hermes 2, consisting of an updated and cleaned version of the OpenHermes 2.5 Dataset, as well as a newly introduced Function Calling and JSON Mode dataset developed in-house.'
    },
    {
        name: 'bge-base-en',
        capability: ['embeddings'],
        provider: 'cloudflare',
        providerModel: '@cf/baai/bge-base-en-v1.5',
        description: 'BAAI general embedding (Base) model that transforms any given text into a 768-dimensional vector'
    },
    {
        name: 'gemma-sea-lion-v4-27b-it',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/aisingapore/gemma-sea-lion-v4-27b-it',
        description: 'SEA-LION stands for Southeast Asian Languages In One Network, which is a collection of Large Language Models (LLMs) which have been pretrained and instruct-tuned for the Southeast Asia (SEA) region.'
    },
    {
        name: 'llama-3.1-8b-instruct-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.1-8b-instruct-awq',
        description: 'Quantized (int4) generative text model with 8 billion parameters from Meta. '
    },
    {
        name: 'uform-gen2-qwen-500m',
        capability: ['vision'],
        provider: 'cloudflare',
        providerModel: '@cf/unum/uform-gen2-qwen-500m',
        description: 'UForm-Gen is a small generative vision-language model primarily designed for Image Captioning and Visual Question Answering. The model was pre-trained on the internal image captioning dataset and fine-tuned on public instructions datasets: SVIT, LVIS, VQAs datasets.'
    },
    {
        name: 'zephyr-7b-beta-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/thebloke/zephyr-7b-beta-awq',
        description: 'Zephyr 7B Beta AWQ is an efficient, accurate and blazing-fast low-bit weight quantized Zephyr model variant.'
    },
    {
        name: 'gemma-7b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/google/gemma-7b-it-lora',
        description: '  This is a Gemma-7B base model that Cloudflare dedicates for inference with LoRA adapters. Gemma is a family of lightweight, state-of-the-art open models from Google, built from the same research and technology used to create the Gemini models.'
    },
    {
        name: 'qwen-1.5-1.8b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/qwen/qwen1.5-1.8b-chat',
        description: 'Qwen1.5 is the improved version of Qwen, the large language model series developed by Alibaba Cloud.'
    },
    {
        name: 'mistral-small-3.1',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/mistralai/mistral-small-3.1-24b-instruct',
        description: 'Building upon Mistral Small 3 (2501), Mistral Small 3.1 (2503) adds state-of-the-art vision understanding and enhances long context capabilities up to 128k tokens without compromising text performance. With 24 billion parameters, this model achieves top-tier capabilities in both text and vision tasks.'
    },
    {
        name: 'llama-3-8b-instruct-awq',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3-8b-instruct-awq',
        description: 'Quantized (int4) generative text model with 8 billion parameters from Meta.'
    },
    {
        name: 'llama-3.2-11b-vision',
        capability: ['chat', 'vision'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.2-11b-vision-instruct',
        description: ' The Llama 3.2-Vision instruction-tuned models are optimized for visual recognition, image reasoning, captioning, and answering general questions about an image.'
    },
    {
        name: 'whisper-tiny',
        capability: ['audio'],
        provider: 'cloudflare',
        providerModel: '@cf/openai/whisper-tiny-en',
        description: 'Whisper is a pre-trained model for automatic speech recognition (ASR) and speech translation. Trained on 680k hours of labelled data, Whisper models demonstrate a strong ability to generalize to many datasets and domains without the need for fine-tuning. This is the English-only version of the Whisper Tiny model which was trained on the task of speech recognition.'
    },
    {
        name: 'aura-1',
        capability: ['tts'],
        provider: 'cloudflare',
        providerModel: '@cf/deepgram/aura-1',
        description: 'Aura is a context-aware text-to-speech (TTS) model that applies natural pacing, expressiveness, and fillers based on the context of the provided text. The quality of your text input directly impacts the naturalness of the audio output.'
    },
    {
        name: 'sqlcoder-7b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/defog/sqlcoder-7b-2',
        description: 'This model is intended to be used by non-technical users to understand data inside their SQL databases. '
    },
    {
        name: 'phi-2',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/microsoft/phi-2',
        description: 'Phi-2 is a Transformer-based model with a next-word prediction objective, trained on 1.4T tokens from multiple passes on a mixture of Synthetic and Web datasets for NLP and coding.'
    },
    {
        name: 'bart-large-cnn',
        capability: ['summarization'],
        provider: 'cloudflare',
        providerModel: '@cf/facebook/bart-large-cnn',
        description: 'BART is a transformer encoder-encoder (seq2seq) model with a bidirectional (BERT-like) encoder and an autoregressive (GPT-like) decoder. You can use this model for text summarization.'
    },
    {
        name: 'stable-diffusion-v1-5-img2img',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/runwayml/stable-diffusion-v1-5-img2img',
        description: 'Stable Diffusion is a latent text-to-image diffusion model capable of generating photo-realistic images. Img2img generate a new image from an input image with Stable Diffusion. '
    },
    {
        name: 'embeddinggemma-300m',
        capability: ['embeddings'],
        provider: 'cloudflare',
        providerModel: '@cf/google/embeddinggemma-300m',
        description: 'EmbeddingGemma is a 300M parameter, state-of-the-art for its size, open embedding model from Google, built from Gemma 3 (with T5Gemma initialization) and the same research and technology used to create Gemini models. EmbeddingGemma produces vector representations of text, making it well-suited for search and retrieval tasks, including classification, clustering, and semantic similarity search. This model was trained with data in 100+ spoken languages.'
    },
    {
        name: 'bge-reranker-base',
        capability: ['text-classification'],
        provider: 'cloudflare',
        providerModel: '@cf/baai/bge-reranker-base',
        description: 'Different from embedding model, reranker uses question and document as input and directly output similarity instead of embedding. You can get a relevance score by inputting query and passage to the reranker. And the score can be mapped to a float value in [0,1] by sigmoid function.  '
    },
    {
        name: 'gemma-7b-it',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@hf/google/gemma-7b-it',
        description: 'Gemma is a family of lightweight, state-of-the-art open models from Google, built from the same research and technology used to create the Gemini models. They are text-to-text, decoder-only large language models, available in English, with open weights, pre-trained variants, and instruction-tuned variants.'
    },
    {
        name: 'lucid-origin',
        capability: ['image-generation'],
        provider: 'cloudflare',
        providerModel: '@cf/leonardo/lucid-origin',
        description: 'Lucid Origin from Leonardo.AI is their most adaptable and prompt-responsive model to date. Whether you\'re generating images with sharp graphic design, stunning full-HD renders, or highly specific creative direction, it adheres closely to your prompts, renders text with accuracy, and supports a wide array of visual styles and aesthetics – from stylized concept art to crisp product mockups. '
    },
    {
        name: 'qwen-1.5-14b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/qwen/qwen1.5-14b-chat-awq',
        description: 'Qwen1.5 is the improved version of Qwen, the large language model series developed by Alibaba Cloud. AWQ is an efficient, accurate and blazing-fast low-bit weight quantization method, currently supporting 4-bit quantization.'
    },
    {
        name: 'openchat-3.5',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/openchat/openchat-3.5-0106',
        description: 'OpenChat is an innovative library of open-source language models, fine-tuned with C-RLFT - a strategy inspired by offline reinforcement learning.'
    },
    {
        name: 'llama-4-scout-17b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-4-scout-17b-16e-instruct',
        description: 'Meta\'s Llama 4 Scout is a 17 billion parameter model with 16 experts that is natively multimodal. These models leverage a mixture-of-experts architecture to offer industry-leading performance in text and image understanding.'
    },
    {
        name: 'gemma-3-12b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/google/gemma-3-12b-it',
        description: 'Gemma 3 models are well-suited for a variety of text generation and image understanding tasks, including question answering, summarization, and reasoning. Gemma 3 models are multimodal, handling text and image input and generating text output, with a large, 128K context window, multilingual support in over 140 languages, and is available in more sizes than previous versions.'
    },
    {
        name: 'qwen-qwq-32b',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/qwen/qwq-32b',
        description: 'QwQ is the reasoning model of the Qwen series. Compared with conventional instruction-tuned models, QwQ, which is capable of thinking and reasoning, can achieve significantly enhanced performance in downstream tasks, especially hard problems. QwQ-32B is the medium-sized reasoning model, which is capable of achieving competitive performance against state-of-the-art reasoning models, e.g., DeepSeek-R1, o1-mini.'
    },
    {
        name: 'bge-large-en',
        capability: ['embeddings'],
        provider: 'cloudflare',
        providerModel: '@cf/baai/bge-large-en-v1.5',
        description: 'BAAI general embedding (Large) model that transforms any given text into a 1024-dimensional vector'
    },
    {
        name: 'aura-2-en',
        capability: ['tts'],
        provider: 'cloudflare',
        providerModel: '@cf/deepgram/aura-2-en',
        description: 'Aura-2 is a context-aware text-to-speech (TTS) model that applies natural pacing, expressiveness, and fillers based on the context of the provided text. The quality of your text input directly impacts the naturalness of the audio output.'
    },
    {
        name: 'llama-3.1-70b-instruct',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.1-70b-instruct',
        description: 'Llama 3.1 70B Instruct'
    },
    {
        name: 'llama-3.1-8b-instruct-fast',
        capability: ['chat'],
        provider: 'cloudflare',
        providerModel: '@cf/meta/llama-3.1-8b-instruct-fast',
        description: 'Llama 3.1 8B Instruct Fast'
    }
];
/** Get all models by capability */
export function getModelsByCapability(capability) {
    return MODEL_CATALOG.filter(model => model.capability.includes(capability));
}
/** Get model information by name */
export function getModelInfo(modelName) {
    return MODEL_CATALOG.find(model => model.name === modelName);
}
;
